app/lib/katello/resources/registry.rb

```diff
@@ -12,22 +12,37 @@ def self.logger
 
         def self.get(path, headers = {:accept => :json})
           logger.debug "Sending GET request to Registry: #{path}"
-          client = RegistryResource.rest_client(Net::HTTP::Get, :get, path)
+          client = RegistryResource.load_class.rest_client(Net::HTTP::Get, :get, path)
           client.get(headers)
         end
       end
 
       class RegistryResource < HttpResource
-        if SETTINGS[:katello][:container_image_registry]
-          cfg = SETTINGS[:katello][:container_image_registry]
-          url = cfg[:crane_url]
-          uri = URI.parse(url)
-          self.prefix = uri.path
-          self.site = "#{uri.scheme}://#{uri.host}:#{uri.port}"
-          self.ca_cert_file = cfg[:crane_ca_cert_file]
-        end
-
         class << self
+          def load_class
+            container_config = SETTINGS.dig(:katello, :container_image_registry)
+            registry_url = nil
+            pulp_master = ::SmartProxy.pulp_master
+
+            # Pulp 3 has its own registry
+            if pulp_master && pulp_master.content_pulp3_support?(::Katello::DockerBlob::CONTENT_TYPE)
+              registry_url = pulp_master.setting('Pulp3', 'content_app_url')
+              # Assume the registry uses the same CA as the Smart Proxy
+              ca_cert_file = Setting[:ssl_ca_file]
+            elsif container_config
+              registry_url = container_config[:crane_url]
+              ca_cert_file = container_config[:registry_ca_cert_file]
+            end
+
+            logger.send(:error, "No container registry url specified") unless registry_url
```

Comment:
Should this be a harder error? `URI.parse(nil)` raises an exception and can happen in misconfigurations.